runtime/lua/vim/filetype.lua

```diff
@@ -69,39 +86,95 @@ end
 ---@param bufnr integer The buffer to get the line from
 ---@param start_lnum integer The line number of the first line to start from (inclusive, 1-based)
 ---@return string|nil The first non-blank line if found or `nil` otherwise
-function M.nextnonblank(bufnr, start_lnum)
-  for _, line in ipairs(M.getlines(bufnr, start_lnum, -1)) do
+function M._nextnonblank(bufnr, start_lnum)
+  for _, line in ipairs(M._getlines(bufnr, start_lnum, -1)) do
     if not line:find('^%s*$') then
       return line
     end
   end
   return nil
 end
 
+--- @type table<string,vim.regex>
+local cache = {}
+
 ---@private
 --- Check whether the given string matches the Vim regex pattern.
-M.matchregex = (function()
-  local cache = {}
-  return function(s, pattern)
-    if s == nil then
-      return nil
+--- @param s string?
+--- @param pattern string
+--- @return boolean?
+function M._matchregex(s, pattern)
+  if not s then
+    return
+  end
+  if not cache[pattern] then
+    cache[pattern] = vim.regex(pattern)
+  end
+  return cache[pattern]:match_str(s) ~= nil
+end
+
+--- @module 'vim.filetype.detect'
+local detect = setmetatable({}, {
+  --- @param k string
+  --- @param t table<string,function>
+  --- @return function
+  __index = function(t, k)
+    t[k] = function(...)
+      return require('vim.filetype.detect')(...)
     end
-    if not cache[pattern] then
-      cache[pattern] = vim.regex(pattern)
+    return t[k]
+  end,
+})
```

Comment:
Can we keep a `do ... end` block around the `local cache` declaration to avoid spilling the (not very specific) variable name into the broader scope?

That was the purpose of the `(function() ... end)()` pattern, which I now regret and wish I had just used `do ... end` to begin with.

Reply:
It's not my favourite style as I find the weight of the additional nesting outweighs the benefits of leaking the variable to other parts of the file. Does it matter if other areas of `filetype.lua` access this cache?

Either way, I'll add the `do-end` back and rename the variables to `regex_cache`.